### PR TITLE
Add events to Google Calendar

### PR DESCRIPTION
**Pull request by natalie-abreu**
- Adds button on each event that allows the user to add it to their calendar
- Saves which events the user has added to calendar so you can't add an event to your calendar multiple times
- Turned UserInfo class into User class (no longer using the Google User class) and adds a User Builder to increase readability/make it easier to use

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/EventsServlet.java [hunk 1]
```diff
@@ -211,4 +259,47 @@ public String sanitizeInput(String input) throws java.io.UnsupportedEncodingExce
       input = new String( input.getBytes("UTF-8") , 0, Math.min(MAX_STRING_BYTES, input.length()), "UTF-8");
       return input;
   }
+
+  public long getEventId() {
```
Comment (gusteran) [inline src/main/java/com/google/sps/servlets/EventsServlet.java:263]: Does this method just return the most recent event added to the database?

File: src/main/java/com/google/sps/servlets/EventsServlet.java [hunk 2]
```diff
@@ -81,18 +92,33 @@
     static final String START_TIME = "start";
     static final String END_TIME = "end";
     static final String USER_TIMEZONE = "timezone";
+    static final String CREATOR = "creator";
+    static final String EVENT_NUM = "EventNum";
+    static final String EVENT_NUM_VALUE = "value";
+    static final String EVENT_ID = "event_id";
+    static final String ID_TOKEN_PARAM = "id_token";
+
+    private static final String CLIENT_ID = "605480199600-e4uo1livbvl58cup3qtd1miqas7vspcu.apps.googleusercontent.com";
```
[the remark this answers is not in the document]
Comment (natalie-abreu) [reply src/main/java/com/google/sps/servlets/EventsServlet.java:101]: Yes, it's part of the OAuth credentials

File: src/main/java/com/google/sps/servlets/EventsServlet.java (hunk 1 above)
Comment (natalie-abreu) [reply src/main/java/com/google/sps/servlets/EventsServlet.java:263]: It tracks how many events have been made in order to sequentially ID the events (first event will get an id of 0, etc)

File: src/main/java/com/google/sps/data/User.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  public static final String ENTITY_KEY = "entity_key";
+
+  private String id;
+  private String nickname;
+  private ArrayList<Long> created_events;
+  private ArrayList<Long> bookmarked_events;
+  private ArrayList<Long> added_to_calendar_events;
+  private Long current_challenge_id;
+  private ArrayList<Integer> challenge_statuses;
+  private Key entity_key;
```
Comment (kavinkarthik92) [inline src/main/java/com/google/sps/data/User.java:44]: Add a comment on what it is and why is it relevant to an user

File: src/main/java/com/google/sps/data/User.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        .setCurrentChallengeId(currentChallengeId)
+        .setChallengeStatuses(challengeStatuses)
+        .build();
+
+    return user;
+  }
+  
+  public Entity toEntity() {
+      Entity userEntity;
+      if (this.entity_key == null) userEntity = new Entity(DATA_TYPE);
```
Comment (kavinkarthik92) [inline src/main/java/com/google/sps/data/User.java:188]: you need to update entity_key here.

If you call toEntity() multiple time, we might end up creating two different Entities with different keys right?

File: src/main/java/com/google/sps/servlets/EventsServlet.java (hunk 2 above)
Comment (kavinkarthik92) [inline src/main/java/com/google/sps/servlets/EventsServlet.java:101]: What is this magic string? Is it unique for each client? Could update documentation on where you them from?

File: src/main/java/com/google/sps/servlets/EventsServlet.java [hunk 5]
```diff
@@ -211,4 +259,47 @@ public String sanitizeInput(String input) throws java.io.UnsupportedEncodingExce
...
+  }
+
+  public void updateUserCreatedEvents(String userId, long eventId) {
+      Query userQuery = new Query(User.DATA_TYPE).setFilter(new FilterPredicate(User.ID, FilterOperator.EQUAL, userId));
+      Entity entity = datastore.prepare(userQuery).asSingleEntity();
+      ArrayList<Long> createdEvents =(ArrayList<Long>) entity.getProperty(User.CREATED_EVENTS);
+      if (createdEvents == null) createdEvents = new ArrayList<Long>();
+      createdEvents.add(eventId);
+      entity.setProperty(User.CREATED_EVENTS, createdEvents);
+      datastore.put(entity);
```
Comment (kavinkarthik92) [inline src/main/java/com/google/sps/servlets/EventsServlet.java:293]: Not fully familiar with datastore. Is there a way to mention the key for a datastore? I seeing that you are inserting a value (entity) in this case. Is there a way to define a PRIMARY KEY for a datastore table?

Comment (natalie-abreu) [reply src/main/java/com/google/sps/servlets/EventsServlet.java:293]: There's a way to set the id which is what I am doing with the entity_key in the User class, but I'm not sure about making another attribute the primary key

File: src/main/java/com/google/sps/data/User.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            this.nickname = nickname;
+            return this; 
+        }
+        public Builder setCreatedEvents(ArrayList<Long> created_events){
+            if (created_events == null) this.created_events = new ArrayList<Long>();
+            else this.created_events = (ArrayList) created_events.clone();
+            return this;
+        }
+        public Builder setBookmarkedEvents(ArrayList<Long> bookmarked_events){
+            if (bookmarked_events == null) this.bookmarked_events = new ArrayList<Long>();
```
[the remark this answers is not in the document]
Comment (natalie-abreu) [reply src/main/java/com/google/sps/data/User.java:73]: Added the null checks to catch when the user entity is pulled from datastore and these attributes are still null -- can check for null values instead in the servlet before constructing a User if you think that makes more sense?

File: src/main/java/com/google/sps/servlets/EventsServlet.java [hunk 7]
```diff
@@ -81,18 +92,33 @@
...
     static final int MAX_STRING_BYTES = 1500;
 
+    static final HttpTransport HTTP_TRANSPORT = new UrlFetchTransport();
+    static final JsonFactory JSON_FACTORY = new GsonFactory();
+
+    static final GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(HTTP_TRANSPORT, JSON_FACTORY)
+        .setAudience(Collections.singletonList(CLIENT_ID))
+        .build();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (lazau-g) [inline src/main/java/com/google/sps/servlets/EventsServlet.java:116]: Why move datastoreService out of each handler and up to here?

File: src/main/java/com/google/sps/servlets/EventsServlet.java [hunk 8]
```diff
@@ -211,4 +259,47 @@ public String sanitizeInput(String input) throws java.io.UnsupportedEncodingExce
...
+      Query userQuery = new Query(User.DATA_TYPE).setFilter(new FilterPredicate(User.ID, FilterOperator.EQUAL, userId));
+      Entity entity = datastore.prepare(userQuery).asSingleEntity();
+      ArrayList<Long> createdEvents =(ArrayList<Long>) entity.getProperty(User.CREATED_EVENTS);
+      if (createdEvents == null) createdEvents = new ArrayList<Long>();
+      createdEvents.add(eventId);
+      entity.setProperty(User.CREATED_EVENTS, createdEvents);
+      datastore.put(entity);
+  }
+
+  private GoogleIdToken verifyId(String idTokenString) {
```
Comment (lazau-g) [inline src/main/java/com/google/sps/servlets/EventsServlet.java:296]: refactor this and UserServlet.java:198 into GoogleIdHelper.java so that both EventsServlet and UsersServlet could use code reading something like:

`Payload p = GoogleIdHelper.verifyId(request)`

The following pattern is seen three times in UserServlet, and twice here - a sure sign it could be in a helper somewhere!
```
GoogleIdToken idToken = verifyId(request.getParameter(ID_TOKEN_PARAM));
    if (idToken == null) {
        System.out.println("Invalid ID token.");
        response.setStatus(400);
        return;
    }

    Payload payload = idToken.getPayload();
```

What the code essentially does is converts the ID_TOKEN request parameter into a usable Payload. This refactor can also make it easier to test by keeping the dependency injection stuff compartmentalized to the GoogleIdHelper.

File: src/main/java/com/google/sps/servlets/EventsServlet.java [hunk 9]
```diff
@@ -106,6 +132,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         Date startTime = (Date) entity.getProperty(START_TIME);
         Date endTime = (Date) entity.getProperty(END_TIME);
         String category = (String) entity.getProperty(CATEGORY);
+        String userId = (String) entity.getProperty(EVENT_CREATOR);
+        long eventId = (long) entity.getProperty(EVENT_ID);
```
Comment (lazau-g) [inline src/main/java/com/google/sps/servlets/EventsServlet.java:136]: Refactor the eventEntity into an sps/data/Events.java class, similar to your Users class.

File: src/main/java/com/google/sps/servlets/EventsServlet.java [hunk 10]
```diff
@@ -81,18 +92,33 @@
     static final String START_TIME = "start";
     static final String END_TIME = "end";
     static final String USER_TIMEZONE = "timezone";
+    static final String EVENT_CREATOR = "creator";
+    static final String EVENT_NUM = "EventNum";
+    static final String EVENT_NUM_VALUE = "value";
+    static final String EVENT_ID = "event_id";
```
Comment (lazau-g) [inline src/main/java/com/google/sps/servlets/EventsServlet.java:98]: Just reuse the datastore assigned Key.getId property for a unique long that can represent an entity. See also my comment on 263 (in response you @gusteran and yourself).

File: src/main/java/com/google/sps/servlets/EventsServlet.java (hunk 1 above)
Comment (lazau-g) [reply src/main/java/com/google/sps/servlets/EventsServlet.java:263]: If you're looking for a unique ID to key events by, `datastore.put(eventEntity)` on line 225 will assign the new eventEntity a Key which has a unique ID which you can use.

By relying on datastore to provide unique IDs for each entity, you can access the unique event id by:
```
datastore.put(newEventEntity);
long myEventUniqueId = newEventEntity.getKey().getId();
```

File: src/main/webapp/index.html [hunk 11]
```diff
@@ -77,6 +81,7 @@ <h1 class="title">🌏project title</h1>
         <p class="close-modal-button" id="close-create-event-modal-button" onclick="closeCreateEventModal()">x</p>
         <form id="create-event-form" action="/events" method="POST">
             <input type="text" id="user-timezone" name="timezone" style="display:none;" required></input>
+            <input type="text" id="id-token" name="id_token" style="display:none;" required></input>
```
Comment (lazau-g) [inline src/main/webapp/index.html:84]: nit: Hidden required nodes can lead to an unpleasant user experience if some JS fails - the user hits submit and the browser highlights a hidden field, which to the end user looks like the browser does nothing and refuses to do anything.

A better approach would be to drop the required attribute, and return an error response in your servlet. In the future, you may chose to surface the error response to the user if you want.

File: src/main/webapp/feed.js [hunk 12]
```diff
@@ -665,20 +683,21 @@ async function updateSigninStatus(isSignedIn) {
...
-    await loadPage();
-    showAddToCalendarButtons();
-} else {
-    authorizeButton.style.display = 'block';
-    signoutButton.style.display = 'none';
-    feedRightSide.style.display = "none";
-    await loadPage();
-    hideAddToCalendarButtons();
-}
+    if (isSignedIn) {
```
Comment (lazau-g) [inline src/main/webapp/feed.js:686]: If the change in this code block is only formatting (re-indentation in this case) - drop it from this PR and make another PR that contains only formatting changes. It makes it simpler to understand blame history.

File: src/main/java/com/google/sps/servlets/EventsServlet.java (hunk 7 above)
Comment (natalie-abreu) [reply src/main/java/com/google/sps/servlets/EventsServlet.java:116]: Previously moved it so I wouldn't have to pass it to each member function that involved datastore, but most of that changed when I moved functionality into the User class so I can move it back to each handler.

File: src/main/java/com/google/sps/servlets/EventsServlet.java (hunk 9 above)
Comment (natalie-abreu) [reply src/main/java/com/google/sps/servlets/EventsServlet.java:136]: Made an EventWrapper class instead of an Event class since I use the Event class from the Google Calendar API -- it's a little messy so let me know if there's a better way of going about it.
